src/shared/components/home/emojis-form.tsx

```diff
@@ -0,0 +1,431 @@
+  }
+
+export class EmojiForm extends Component<any, EmojiFormState> {
+    private isoData = setIsoData(this.context);
+    private subscription: Subscription | undefined;
+    private emptyState: EmojiFormState = {
+        loading: false,
+        siteRes: this.isoData.site_res,
+        custom_emojis: this.isoData.site_res.custom_emojis.map(x =>
+        ({
+            id: x.custom_emoji.id,
+            category: x.custom_emoji.category,
+            shortcode: x.custom_emoji.shortcode,
+            image_url: x.custom_emoji.image_url,
+            alt_text: x.custom_emoji.alt_text,
+            keywords: x.keywords.map(x => x.keyword).join(" "),
+            changed: false,
+        }))
+    };
+    private emojiLookup = {};
+    private scrollRef: any = {};
+    constructor(props: any, context: any) {
+        super(props, context);
+        this.state = this.emptyState;
+
+        this.parseMessage = this.parseMessage.bind(this);
+        this.subscription = wsSubscribe(this.parseMessage);
+
+        this.emojiLookup = getCustomEmojiLookup();
+    }
+    get documentTitle(): string {
+        return i18n.t("custom_emojis");
+    }
+
+    componentWillUnmount() {
+        if (isBrowser()) {
+          this.subscription?.unsubscribe();
+        }
+      }
+
+    render() {
+        return (
+            <div className="col-12">
+                <Prompt
+                    when={!this.state!.loading}
+                    message={i18n.t("block_leaving")}
+                />
+                <HtmlTags
+                    title={this.documentTitle}
+                    path={this.context.router.route.match.url}
+                />
+                <h5 className="col-12">{i18n.t("custom_emojis")}</h5>
+                {Object.keys(this.emojiLookup).length > 0 && <div>
+                    <EmojiMart onEmojiClick={this.handleEmojiClick} pickerOptions={this.configurePicker()}></EmojiMart>
+                </div>}
+                <div className="table-responsive col-12">
+                    <table
+                        id="emojis_table"
+                        className="table table-hover"
+                    >
+                        <thead className="pointer">
+                            <tr>
+                                <th style="width:70px;">{i18n.t("column_emoji")}</th>
+                                <th style="width:150px">{i18n.t("column_shortcode")}</th>
+                                <th style="width:150px">{i18n.t("column_category")}</th>
+                                <th style="width:200px">{i18n.t("column_imageurl")}</th>
+                                <th style="width:100px">{i18n.t("column_alttext")}</th>
+                                <th>{i18n.t("column_keywords")}</th>
+                                <th style="width:107px"></th>
+                            </tr>
+                        </thead>
+                        <tbody>
+                            {this.state!.custom_emojis
```

Comment:
I think these exclamations can cause crashes right? State should already be initialized, so I'm not sure why the IDE isn't telling you this is redundant.